<a href="https://colab.research.google.com/github/michellssGUi/dataScience/blob/main/Wallmart_Melhor_Loja.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Análise Lojas Wallmart - 23/05/2022

---

# Introdução

## 1.1 - Objetivo do Projeto
> Fazer um levantamento do faturamento das lojas nos USA e apontar qual loja seria melhor para expandir seu tamaho.

## 1.2 Descrevendo os dados:
> Os dados apresentam as vendas semanais de 2010/02/05 à 2012/11/12 em umtotal de 45 lojas varejistas da rede Wallmart e também algumas métricas econômicas e meteorológicas por semana. 

A descrição destas características estão abaixo:

## Feriados Importantes

Holiday | Semanas
----|----
Super Bowl |	12-Feb-10, 11-Feb-11, 10-Feb-12
Labour Day | 10-Sep-10, 09-Sep-11, 07-Sep-12
Thanksgiving | 26-Nov-10, 25-Nov-11, 23-Nov-12
Cristmans | 31-Dec-10, 30-Dec-11, 28-Dec-12

## Index's das tabelas e suas descrições

Variável | Descrição
----|----
Store | Número da loja
Date | Semana de venda
Weekly_Sales | Vendas Nauela Semana
Hliday_Flag | Flag se é ou não semana com feriado (1 - Holiday Weeh e 0 - Non Holiday Week)
Temeratura | Temperatura do dia em °F
Fuel_Price | Preço do combustível na região da loja
CPI | Índice de preços ao consumidor
Unemployment | Taxa de desemprego


## 2 - Os dados
> Realizandoimportação das bibliotecas python as quais serão utilizadas durante o processo de limpeza, organização e análise dos dados.

In [540]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import scipy as scp
import seaborn as sns

### 2.1 - Importanto os dados
> Os dados se encontram em um aquivo .CSV e foram importados fazendo us da biplioteca "pandas".

In [541]:
df = pd.read_csv('Walmart.csv')
df

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0         1  05-02-2010    1643690.90             0        42.31       2.572   
1         1  12-02-2010    1641957.44             1        38.51       2.548   
2         1  19-02-2010    1611968.17             0        39.93       2.514   
3         1  26-02-2010    1409727.59             0        46.63       2.561   
4         1  05-03-2010    1554806.68             0        46.50       2.625   
...     ...         ...           ...           ...          ...         ...   
6430     45  28-09-2012     713173.95             0        64.88       3.997   
6431     45  05-10-2012     733455.07             0        64.89       3.985   
6432     45  12-10-2012     734464.36             0        54.47       4.000   
6433     45  19-10-2012     718125.53             0        56.47       3.969   
6434     45  26-10-2012     760281.43             0        58.85       3.882   

             CPI  Unemployment  
0     211.096358         8.106  
1     211.242170         8.106  
2     211.289143         8.106  
3     211.319643         8.106  
4     211.350143         8.106  
...          ...           ...  
6430  192.013558         8.684  
6431  192.170412         8.667  
6432  192.327265         8.667  
6433  192.330854         8.667  
6434  192.308899         8.667  

[6435 rows x 8 columns]

In [542]:
df = df.sort_values("Weekly_Sales")
df

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
4619     33  03-12-2010     209986.25             0        52.82       3.041   
4614     33  29-10-2010     213538.32             0        71.34       3.130   
4675     33  30-12-2011     215359.21             1        51.60       3.428   
4623     33  31-12-2010     219804.85             1        52.91       3.148   
4671     33  02-12-2011     220060.35             0        59.12       3.701   
...     ...         ...           ...           ...          ...         ...   
1762     13  24-12-2010    3595903.20             0        34.90       2.846   
527       4  23-12-2011    3676388.98             0        35.92       3.103   
1333     10  24-12-2010    3749057.69             0        57.06       3.236   
2763     20  24-12-2010    3766687.43             0        25.17       3.141   
1905     14  24-12-2010    3818686.45             0        30.59       3.141   

             CPI  Unemployment  
4619  126.731333         9.265  
4614  126.436419         9.265  
4675  130.071032         8.010  
4623  127.087677         9.265  
4671  129.845967         8.010  
...          ...           ...  
1762  126.983581         7.795  
527   129.984548         5.143  
1333  126.983581         9.003  
2763  204.637673         7.484  
1905  182.544590         8.724  

[6435 rows x 8 columns]

## 2.2 - Limpeza e preparação dos dados
> Durante esse processo preparandos os dados para análise realizando a limpeza, transforamção e reorganização à medida que for necessário.

### 2.2.1 - Tratamento dos dados ausentes
> Dados ausentes são comuns em muitas aplicações de análise de dados. Um dos objetivos desse passo é deixar o trabalho o menos prolemático possível.


> Por padrão, "dropna()" descarta qualquer linha contendo um valor ausente. Devido a isso é importante avaliar se a perda dessa linhas não irá impactar diretamento na análise dos dados.

In [543]:
df = df.dropna()
df

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
4619     33  03-12-2010     209986.25             0        52.82       3.041   
4614     33  29-10-2010     213538.32             0        71.34       3.130   
4675     33  30-12-2011     215359.21             1        51.60       3.428   
4623     33  31-12-2010     219804.85             1        52.91       3.148   
4671     33  02-12-2011     220060.35             0        59.12       3.701   
...     ...         ...           ...           ...          ...         ...   
1762     13  24-12-2010    3595903.20             0        34.90       2.846   
527       4  23-12-2011    3676388.98             0        35.92       3.103   
1333     10  24-12-2010    3749057.69             0        57.06       3.236   
2763     20  24-12-2010    3766687.43             0        25.17       3.141   
1905     14  24-12-2010    3818686.45             0        30.59       3.141   

             CPI  Unemployment  
4619  126.731333         9.265  
4614  126.436419         9.265  
4675  130.071032         8.010  
4623  127.087677         9.265  
4671  129.845967         8.010  
...          ...           ...  
1762  126.983581         7.795  
527   129.984548         5.143  
1333  126.983581         9.003  
2763  204.637673         7.484  
1905  182.544590         8.724  

[6435 rows x 8 columns]

> Nesse caso não hove perda de lihas, continuamos com as mesmas 6435 row's e 8 colums.

> OBS: Em caso de se perceber a perda demasiada de row's e por conseuência a perda de dados importantes, poderia-se fazer a substituição dessesdados por um alor consntante, pela média e ou mediana dos demais valores.

Segue exemplo:

> A funçao "fillna()" devolve um novo objeto reindexado com as condições que foram passadas... nesse caso retorna a media cos campos de valor 'NaN'.

In [544]:
df = df.fillna(df.mean())
df

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
4619     33  03-12-2010     209986.25             0        52.82       3.041   
4614     33  29-10-2010     213538.32             0        71.34       3.130   
4675     33  30-12-2011     215359.21             1        51.60       3.428   
4623     33  31-12-2010     219804.85             1        52.91       3.148   
4671     33  02-12-2011     220060.35             0        59.12       3.701   
...     ...         ...           ...           ...          ...         ...   
1762     13  24-12-2010    3595903.20             0        34.90       2.846   
527       4  23-12-2011    3676388.98             0        35.92       3.103   
1333     10  24-12-2010    3749057.69             0        57.06       3.236   
2763     20  24-12-2010    3766687.43             0        25.17       3.141   
1905     14  24-12-2010    3818686.45             0        30.59       3.141   

             CPI  Unemployment  
4619  126.731333         9.265  
4614  126.436419         9.265  
4675  130.071032         8.010  
4623  127.087677         9.265  
4671  129.845967         8.010  
...          ...           ...  
1762  126.983581         7.795  
527   129.984548         5.143  
1333  126.983581         9.003  
2763  204.637673         7.484  
1905  182.544590         8.724  

[6435 rows x 8 columns]

### 2.2.2 - Tratamento de dados duplicados
> Linhas duplicadaspodem ser encontradas em um DataFrame pr diversos motivos, por isso se faz necessário verificar sua existência e tata-los.

> Farei uso do método "duplicated()" para verificar a existência de dados dupicados.

In [545]:
df.duplicated()

4619    False
4614    False
4675    False
4623    False
4671    False
        ...  
1762    False
527     False
1333    False
2763    False
1905    False
Length: 6435, dtype: bool

In [546]:
df = df.drop_duplicates()
pd.DataFrame(df)

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
4619     33  03-12-2010     209986.25             0        52.82       3.041   
4614     33  29-10-2010     213538.32             0        71.34       3.130   
4675     33  30-12-2011     215359.21             1        51.60       3.428   
4623     33  31-12-2010     219804.85             1        52.91       3.148   
4671     33  02-12-2011     220060.35             0        59.12       3.701   
...     ...         ...           ...           ...          ...         ...   
1762     13  24-12-2010    3595903.20             0        34.90       2.846   
527       4  23-12-2011    3676388.98             0        35.92       3.103   
1333     10  24-12-2010    3749057.69             0        57.06       3.236   
2763     20  24-12-2010    3766687.43             0        25.17       3.141   
1905     14  24-12-2010    3818686.45             0        30.59       3.141   

             CPI  Unemployment  
4619  126.731333         9.265  
4614  126.436419         9.265  
4675  130.071032         8.010  
4623  127.087677         9.265  
4671  129.845967         8.010  
...          ...           ...  
1762  126.983581         7.795  
527   129.984548         5.143  
1333  126.983581         9.003  
2763  204.637673         7.484  
1905  182.544590         8.724  

[6435 rows x 8 columns]

### 2.2.3 - Renomeando os índices
> Para maior facilidade na maniulação dos dados estarei passando os indices para letras miusculas.

In [547]:
df.rename(columns=str.lower, inplace=True)
pd.DataFrame.head(df)

store        date  weekly_sales  holiday_flag  temperature  fuel_price  \
4619     33  03-12-2010     209986.25             0        52.82       3.041   
4614     33  29-10-2010     213538.32             0        71.34       3.130   
4675     33  30-12-2011     215359.21             1        51.60       3.428   
4623     33  31-12-2010     219804.85             1        52.91       3.148   
4671     33  02-12-2011     220060.35             0        59.12       3.701   

             cpi  unemployment  
4619  126.731333         9.265  
4614  126.436419         9.265  
4675  130.071032         8.010  
4623  127.087677         9.265  
4671  129.845967         8.010

## 3 - Análise dos dados

### 3.1 - Qual o preço médio de vendas semanal por loja?

In [548]:
df.describe(include='all') 

store        date  weekly_sales  holiday_flag  temperature  \
count   6435.000000        6435  6.435000e+03   6435.000000  6435.000000   
unique          NaN         143           NaN           NaN          NaN   
top             NaN  03-12-2010           NaN           NaN          NaN   
freq            NaN          45           NaN           NaN          NaN   
mean      23.000000         NaN  1.046965e+06      0.069930    60.663782   
std       12.988182         NaN  5.643666e+05      0.255049    18.444933   
min        1.000000         NaN  2.099862e+05      0.000000    -2.060000   
25%       12.000000         NaN  5.533501e+05      0.000000    47.460000   
50%       23.000000         NaN  9.607460e+05      0.000000    62.670000   
75%       34.000000         NaN  1.420159e+06      0.000000    74.940000   
max       45.000000         NaN  3.818686e+06      1.000000   100.140000   

         fuel_price          cpi  unemployment  
count   6435.000000  6435.000000   6435.000000  
unique          NaN          NaN           NaN  
top             NaN          NaN           NaN  
freq            NaN          NaN           NaN  
mean       3.358607   171.578394      7.999151  
std        0.459020    39.356712      1.875885  
min        2.472000   126.064000      3.879000  
25%        2.933000   131.735000      6.891000  
50%        3.445000   182.616521      7.874000  
75%        3.735000   212.743293      8.622000  
max        4.468000   227.232807     14.313000

In [549]:
total_lojas = 45
media_weekly = []
dicionario_media_weekly = {}
for i in range(total_lojas):
  venda = []
  for indice, linha in df.iterrows():
    if linha['store'] == i + 1:
      venda.append(linha['weekly_sales'])
  venda_soma = sum(venda)
  media_loja = venda_soma / 143
  media_weekly.append(media_loja)
  dicionario_media_weekly[i + 1] = dicionario_media_weekly.setdefault(linha['store'] == i + 1, 0) + venda_soma
  #loja_media_vendas = {i + 1 : media_loja}
  print(f'Média das vendas semanais loja_{i + 1}: {media_loja:.2f}') 

Média das vendas semanais loja_1: 1555264.40
Média das vendas semanais loja_2: 1925751.34
Média das vendas semanais loja_3: 402704.44
Média das vendas semanais loja_4: 2094712.96
Média das vendas semanais loja_5: 318011.81
Média das vendas semanais loja_6: 1564728.19
Média das vendas semanais loja_7: 570617.31
Média das vendas semanais loja_8: 908749.52
Média das vendas semanais loja_9: 543980.55
Média das vendas semanais loja_10: 1899424.57
Média das vendas semanais loja_11: 1356383.12
Média das vendas semanais loja_12: 1009001.61
Média das vendas semanais loja_13: 2003620.31
Média das vendas semanais loja_14: 2020978.40
Média das vendas semanais loja_15: 623312.47
Média das vendas semanais loja_16: 519247.73
Média das vendas semanais loja_17: 893581.39
Média das vendas semanais loja_18: 1084718.42
Média das vendas semanais loja_19: 1444999.04
Média das vendas semanais loja_20: 2107676.87
Média das vendas semanais loja_21: 756069.08
Média das vendas semanais loja_22: 1028501.04
Média 

In [550]:
#Loja com maior veda no total
for loja, vendas in dicionario_media_weekly.items():
     print(f"loja {loja}: R${vendas:0.02f}")

loja False: R$0.00
loja 1: R$222402808.85
loja 2: R$275382440.98
loja 3: R$57586735.07
loja 4: R$299543953.38
loja 5: R$45475688.90
loja 6: R$223756130.64
loja 7: R$81598275.14
loja 8: R$129951181.13
loja 9: R$77789218.99
loja 10: R$271617713.89
loja 11: R$193962786.80
loja 12: R$144287230.15
loja 13: R$286517703.80
loja 14: R$511402720.19
loja 15: R$89133683.92
loja 16: R$74252425.40
loja 17: R$127782138.83
loja 18: R$155114734.21
loja 19: R$206634862.10
loja 20: R$301397792.46
loja 21: R$108117878.92
loja 22: R$147075648.57
loja 23: R$198750617.85
loja 24: R$194016021.28
loja 25: R$101061179.17
loja 26: R$143416393.79
loja 27: R$253855916.88
loja 28: R$189263680.58
loja 29: R$77141554.31
loja 30: R$62716885.12
loja 31: R$199613905.50
loja 32: R$166819246.16
loja 33: R$37160221.96
loja 34: R$138249763.00
loja 35: R$131520672.08
loja 36: R$53412214.97
loja 37: R$74202740.32
loja 38: R$55159626.42
loja 39: R$207445542.47
loja 40: R$137870309.79
loja 41: R$181341934.89
loja 42: R$7956575

In [552]:
df_lista_lojas_mais = pd.concat({k: pd.Series(v) for k, v in dicionario_media_weekly.items()}).reset_index()
df_lista_lojas_mais.columns = ['loja', 'drop_all','valor_vendas']
print(df_lista_lojas_mais)

     loja  drop_all  valor_vendas
0   False         0  0.000000e+00
1       1         0  2.224028e+08
2       2         0  2.753824e+08
3       3         0  5.758674e+07
4       4         0  2.995440e+08
5       5         0  4.547569e+07
6       6         0  2.237561e+08
7       7         0  8.159828e+07
8       8         0  1.299512e+08
9       9         0  7.778922e+07
10     10         0  2.716177e+08
11     11         0  1.939628e+08
12     12         0  1.442872e+08
13     13         0  2.865177e+08
14     14         0  5.114027e+08
15     15         0  8.913368e+07
16     16         0  7.425243e+07
17     17         0  1.277821e+08
18     18         0  1.551147e+08
19     19         0  2.066349e+08
20     20         0  3.013978e+08
21     21         0  1.081179e+08
22     22         0  1.470756e+08
23     23         0  1.987506e+08
24     24         0  1.940160e+08
25     25         0  1.010612e+08
26     26         0  1.434164e+08
27     27         0  2.538559e+08
28     28     

In [573]:
df_lista_lojas_mais = pd.DataFrame.head(df_lista_lojas_mais)
df_lista_lojas_mais

0             1
19  20  2.107677e+06
3    4  2.094713e+06
13  14  2.020978e+06
12  13  2.003620e+06
1    2  1.925751e+06

> Lojas que mais venderam

In [576]:
df_lista_lojas_mais = [20, 4, 14, 13, 2]
df_lista_lojas_mais

[20, 4, 14, 13, 2]

### 3.2 - Qual o atributo que tem mais correlação (positiva ou negativa) com o numero de vendas? (E que faça sentido)

In [ ]:
df

### 3.3 - Da loja com maior venda acumulada (soma de vendas de todo o periodo), quantas semanas do ano ele ultrapassa a média do período?

In [ ]:
df_loja = df[['store']] == 21

### 3.4 - Fazendo um gráfico com as vendas máximas, mínimas e médias de todas as lojas ao longo do tempo. Ressaltando o período de feriados mencionados acima.

### 3.5 - Loja Escolhida 